akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/SqlServerOffsetStoreDao.scala

```diff
@@ -0,0 +1,127 @@
+
+  override protected implicit def queryAdapter: QueryAdapter = SqlServerQueryAdapter
+  override protected implicit def timestampCodec: TimestampCodec = SqlServerTimestampCodec
+
+  override protected def createUpsertOffsetSql() = sql"""
+            UPDATE $offsetTable SET
+              current_offset = @currentOffset,
+              manifest = @manifest,
+              mergeable = @mergeable,
+              last_updated = @lastUpdated
+            WHERE projection_name = @projectionName AND projection_key = @projectionKey
+            if @@ROWCOUNT = 0
+              INSERT INTO $offsetTable
+                (projection_name, projection_key, current_offset, manifest, mergeable, last_updated)
+                VALUES (@projectionName, @projectionKey, @currentOffset, @manifest, @mergeable, @lastUpdated)
+              """
+
+  override protected def bindUpsertOffsetSql(
+      stmt: Statement,
+      singleOffset: SingleOffset,
+      toEpochMilli: Long): Statement = {
+    stmt
+      .bind("@projectionName", singleOffset.id.name)
+      .bind("@projectionKey", singleOffset.id.key)
+      .bind("@currentOffset", singleOffset.offsetStr)
+      .bind("@manifest", singleOffset.manifest)
+      .bind("@mergeable", java.lang.Boolean.valueOf(singleOffset.mergeable))
+      .bind("@lastUpdated", toEpochMilli)
+  }
+
+  override protected def createUpdateManagementStateSql(): String = {
+    sql"""
+         UPDATE $managementTable SET
+           paused = @paused,
+           last_updated = @lastUpdated
+         WHERE projection_name = @projectionName AND projection_key = @projectionKey
+         if @@ROWCOUNT = 0
+           INSERT INTO $managementTable
+           (projection_name, projection_key, paused, last_updated)
+           VALUES (@projectionName, @projectionKey, @paused, @lastUpdated)
+         """
+  }
+
+  override protected def bindUpdateManagementStateSql(
+      stmt: Statement,
+      projectionId: ProjectionId,
+      paused: Boolean,
+      lastUpdated: Long): Statement = {
+    stmt
+      .bind("@paused", paused)
+      .bind("@lastUpdated", lastUpdated)
+      .bind("@projectionName", projectionId.name)
+      .bind("@projectionKey", projectionId.key)
+  }
+
+  /**
+   * The r2dbc-sqlserver driver seems to not support binding of array[T].
+   * So have to bake the param into the statement instead of binding it.
+   *
+   * @param notInLatestBySlice not used in postgres, but needed in sql
+   * @return
+   */
+  override protected def deleteOldTimestampOffsetSql(notInLatestBySlice: Seq[String]): String = {
+    val base =
+      s"DELETE FROM $timestampOffsetTable WHERE slice BETWEEN ? AND ? AND projection_name = ? AND timestamp_offset < ?"
+    if (notInLatestBySlice.isEmpty) {
+      sql"$base"
+    } else {
+      val listParam = notInLatestBySlice
+        .map(item => s"'$item'")
+        .mkString(",")
+      sql"""
+      $base
+      AND CONCAT(persistence_id, '-', seq_nr) NOT IN ($listParam)"""
```

Comment:
not possible to use a bind parameter for that `$listParam`? such things could be source of sql injection

Reply:
I don't think so,  but I can have another look.

Reply:
No, does not seem it works:
```
Cause: java.lang.IllegalArgumentException: Cannot encode [[Ljava.lang.String;@5b565160] parameter of type [[Ljava.lang.String;]
```

The  currently supported codecs can be found [here](https://github.com/r2dbc/r2dbc-mssql/blob/main/src/main/java/io/r2dbc/mssql/codec/DefaultCodecs.java#L53-L75) representing [this list of java types](https://github.com/r2dbc/r2dbc-mssql/blob/main/src/main/java/io/r2dbc/mssql/message/type/SqlServerType.java#L39-L75). The maintainer does mention this as well [here](https://github.com/r2dbc/r2dbc-mssql/issues/210#issuecomment-871366198).

I am trying now to transform each item within `notInLatestBySlice` to its own param to end up with `... NOT IN(?,?,?,...)` and then one single `.bind()` per element.

Reply:
I my last commit I removed the `pending` case for sqlserver. The test is still flaky, but at least it went green now to indicate this change ⬆️ works. It if keeps being flaky, we can re-add the `pending` exclusion, but not we at least know it works.

Reply:
Good that you found a way to use parameters.

Now the concat of the pid and seqNr in `notInLatestBySlice` is done in the OffsetStore, and then the dao is assumed to use the same format. We can make a more clear interface by adding a:
```
case class LatestBySlice(slice: Int, pid: String, seqNr: Long)
```
and the dao method would be 
```
def deleteOldTimestampOffsetSql(notInLatestBySlice: Seq[LatestBySlice])
```

The main reason why I started thinking about that was that it might be good to use  fixed number of parameters in the sql. Otherwise we can end up with very many different prepared statements.

That fixed number of parameters would then be the same as number of slices in the slice range. For the slices that are not included in the `notInLatestBySlice` we could still bind a fake value (such as "-") to the parameter.

Reply:
Maybe I can fix this, let me see.

Reply:
That is right, should be at most 1024 parameters because we have at most 1024 slices for one Projection instance (for one R2dbcOffsetStore).

Reply:
I pushed my take on how I understood your suggestion.